### PR TITLE
feat: limit namespace import identifier in id-length rule

### PR DESCRIPTION
**Pull request by Chaedie**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

limit namespace import identifier in id-length rule. 
I did it because of the Issue #18847 

#### Is there anything you'd like reviewers to focus on?

No i don't.

This is my first open-source activity, If i missed something notice me please. 
Thank you for all of your effort in eslint. 
Thanks

### CODE REVIEW

File: tests/lib/rules/id-length.js
[the remark this answers is not in the document]
Comment (Chaedie) [reply tests/lib/rules/id-length.js:1]: @JoshuaKGoldberg 
Thank you for advice. I added some valid cases in [HERE](https://github.com/eslint/eslint/pull/18849/commits/06cb150625345f0e844c93d359a91ed0328fedb2#)
Please let me know If i missed something.

**Comment by nzakas** [comment]
The issue has been accepted so moving out of "Blocked"

File: lib/rules/id-length.js [hunk 1]
```diff
@@ -117,6 +117,8 @@ module.exports = {
                 return properties && !parent.computed && parent.key.name === node.name;
             },
             ImportDefaultSpecifier: true,
+            ImportSpecifier: true,
```
Comment (Tanujkanti4441) [inline lib/rules/id-length.js:120]: I think `ImportSpecifier` should not be reported by the rule but the name assigned to it should.

```js
/*eslint id-length: ["error"]*/

import { x } from 'module';  // rule will not report 'x'

import { x as z } from 'module'; // rule will report report 'z'

```

File: tests/lib/rules/id-length.js [hunk 2]
```diff
@@ -246,6 +248,27 @@ ruleTester.run("id-length", rule, {
         { code: "var [i] = arr;", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
         { code: "var [,i,a] = arr;", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError, tooShortError] },
         { code: "function foo([a]) {}", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
+        { code: "import x from 'module';", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
+        { code: "import * as x from 'module';", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
+        { code: "import { x } from 'module';", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
```
Comment (Tanujkanti4441) [inline tests/lib/rules/id-length.js:253]: This must be a valid test case as explained above.

File: tests/lib/rules/id-length.js [hunk 3]
```diff
@@ -246,6 +248,27 @@ ruleTester.run("id-length", rule, {
...
+            options: [{ max: 5 }],
+            languageOptions: { ecmaVersion: 6 },
+            errors: [tooLongError]
+        },
+        {
+            code: "import { longName } from 'module';",
+            options: [{ max: 5 }],
+            languageOptions: { ecmaVersion: 6 },
+            errors: [tooLongError]
+        },
```
Comment (Tanujkanti4441) [inline tests/lib/rules/id-length.js:271]: This also should be a valid case. Can you update this also?

File: lib/rules/id-length.js (hunk 1 above)
Comment (mdjermanovic) [reply lib/rules/id-length.js:120]: Let's keep the scope of this change limited to the accepted issue https://github.com/eslint/eslint/issues/18847. That is, only check `ImportNamespaceSpecifier`.

Comment (Tanujkanti4441) [reply lib/rules/id-length.js:120]: Make sense to me! for this we can create an another issue, let me re-write the suggestions.

File: lib/rules/id-length.js [hunk 4]
```diff
@@ -117,6 +117,8 @@ module.exports = {
...
+            ImportNamespaceSpecifier: true,
             RestElement: true,
```
Comment (Tanujkanti4441) [inline lib/rules/id-length.js:122]:
```suggestion
            ImportDefaultSpecifier: true,
            ImportNamespaceSpecifier: true,
            RestElement: true,
```

as proposed in the issue we only gonna report `ImportNamespaceSpecifier`

File: tests/lib/rules/id-length.js (hunk 2 above)
Comment (Tanujkanti4441) [inline tests/lib/rules/id-length.js:253]:
```suggestion
        { code: "import x from 'module';", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
        { code: "import * as x from 'module';", languageOptions: { ecmaVersion: 6 }, errors: [tooShortError] },
```

File: tests/lib/rules/id-length.js (hunk 3 above)
Comment (Tanujkanti4441) [inline tests/lib/rules/id-length.js:271]:
```suggestion
        {
            code: "import * as longName from 'module';",
            options: [{ max: 5 }],
            languageOptions: { ecmaVersion: 6 },
            errors: [tooLongError]
        },
```

**Review by Tanujkanti4441 — CHANGES_REQUESTED** [review]

**Comment by Chaedie** [comment]
@Tanujkanti4441 
Thank you for advice, I accepted your suggestions, please check this out.

File: tests/lib/rules/id-length.js [hunk 5]
```diff
@@ -64,6 +64,8 @@ ruleTester.run("id-length", rule, {
         { code: "({ a: obj.x.y.z } = {});", options: [{ properties: "never" }], languageOptions: { ecmaVersion: 6 } },
         { code: "import something from 'y';", languageOptions: { ecmaVersion: 6, sourceType: "module" } },
         { code: "export var num = 0;", languageOptions: { ecmaVersion: 6, sourceType: "module" } },
+        { code: "import * as something from 'y';", languageOptions: { ecmaVersion: 6, sourceType: "module" } },
+        { code: "import { something } from 'y';", languageOptions: { ecmaVersion: 6, sourceType: "module" } },
```
Comment (Tanujkanti4441) [inline tests/lib/rules/id-length.js:68]:
```suggestion
        { code: "import { x } from 'y';", languageOptions: { ecmaVersion: 6, sourceType: "module" } },
```

It shows that `ImportSpecifiers` are allowed even with default `2`.

**Review by Tanujkanti4441 — CHANGES_REQUESTED** [review]

File: tests/lib/rules/id-length.js (hunk 5 above)
Comment (Chaedie) [reply tests/lib/rules/id-length.js]: Good point. 
I think we should add opposite case like
```javascript
{ code: "import { longName } from 'y';", options: [{ max: 5 }], languageOptions: { ecmaVersion: 6, sourceType: "module" } },```

**Review by Tanujkanti4441 — APPROVED** [review]
LGTM, thanks for contributing!

**Review by JoshuaKGoldberg — APPROVED** [review]
🚀

**Pull request merged**